rpc: make rpc.Context configurable

andreimatei opened the pull request:
The rpc.Context does several things:
1) Acts as a utility for opening and caching gRPC connections.
2) Performs periodic "heartbeats" on the open connections and maintains
a health status for each remote address. That status can be queries and
is used to de-prioritize unhealthy replicas.
3) Performs periodic clock skew detection, piggybacking clock signals on
the heartbeats.

Not all clients of the rpc.Context need 2 and 3. In particular, RPC
clients that are "outside of a cluster" (i.e. are not nodes) don't need
to participate in clock skew detection and may or may not want to do the
heartbeats. The fact that they were always performing clock detection
was leading to corrupted state in the RemoteClockMonitor, which was
recording clock info for an "unresolved" remote address (port=0) for
RPC clients that had not been configured with an address. This patch
asserts that we don't do that any more.

The patch makes everything that the rpc.Context does configurable and
uses proper configurations in tests.

andreimatei committed:
rpc: make rpc.Context configurable

The rpc.Context does several things:
1) Acts as a utility for opening and caching gRPC connections.
2) Performs periodic "heartbeats" on the open connections and maintains
a health status for each remote address. That status can be queries and
is used to de-prioritize unhealthy replicas.
3) Performs periodic clock skew detection, piggybacking clock signals on
the heartbeats.

Not all clients of the rpc.Context need 2 and 3. In particular, RPC
clients that are "outside of a cluster" (i.e. are not nodes) don't need
to participate in clock skew detection and may or may not want to do the
heartbeats. The fact that they were always performing clock detection
was leading to corrupted state in the RemoteClockMonitor, which was
recording clock info for an "unresolved" remote address (port=0) for
RPC clients that had not been configured with an address. This patch
asserts that we don't do that any more.

The patch makes everything that the rpc.Context does configurable and
uses proper configurations in tests.

RaduBerinde commented on pkg/rpc/context.go:
```diff
@@ -239,6 +306,18 @@ func (ctx *Context) GRPCDial(target string, opts ...grpc.DialOption) (*grpc.Clie
+				// Send periodic pings on the connection.
+				Time: ctx.cfg.KeepaliveInterval,
+				// If the pings don't get a response within the timeout, the connection
+				// will be closed: if we got a timeout, we might be experiencing a
```
remove "if we got a timeout" (it makes this harder to read)

RaduBerinde commented on pkg/rpc/context.go:
```diff
@@ -107,52 +108,118 @@ type connMeta struct {
+	// RPCs will receive an error. gRPC will attempt to reconnect.
+	KeepaliveTimeout time.Duration
+
+	// HeartbeatInterval is the period on which the health status of a connection
```
"health status of a connection" suggests a lot of overlap with the keepalive. Maybe more accurate to say "the health status of a remote node" instead?

RaduBerinde commented on pkg/testutils/testcluster/testcluster_test.go:
```diff
@@ -211,8 +211,16 @@ func TestStopServer(t *testing.T) {
+	}
 	rpcContext := rpc.NewContext(
-		log.AmbientContext{}, tc.Server(1).RPCContext().Config, tc.Server(1).Clock(), tc.Stopper(),
+		log.AmbientContext{}, rpcCfg, tc.Stopper(),
```
[nit] this fits in one line now

RaduBerinde commented on pkg/rpc/context.go:
```diff
@@ -107,52 +108,118 @@ type connMeta struct {
+		ctx.cfg.KeepaliveInterval = defaultKeepaliveInterval
+	}
+	if ctx.cfg.KeepaliveTimeout == 0 {
+		ctx.cfg.KeepaliveTimeout = 2 * defaultKeepaliveInterval
```
this value should be a constant next to defaultKeepaliveInterval

RaduBerinde reviewed (commented):
Nice cleanup! (though it would have been nice if the cleanup was in a separate commit from the keepalive stuff 😃)

asubiotto commented:
Review status: 0 of 38 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/rpc/context_test.go, line 594 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgF9Vkoej6HamuSX7Tc:-KgF9VkpN-6PQM-WuGEp:b6ol4qo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context_test.go#L594)):*
> ```Go
> 	// One node is expected to see all the other nodes as having too much skew.
> 	// All the other nodes are expected to see that one node as having a large
> 	// skew, and thus pass verification (we verify that a mojority of nodes are
> ```

s/mojority/majority

---

*[pkg/server/server_test.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFACB-4r9JSzDY39O9:-KgFACB-4r9JSzDY39OA:belelkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/server/server_test.go#L121)):*
> ```Go
> 	})
> 
> 	// Edit the server's config so that it's goint to report its admin url using
> ```

s/it's goint to report/it reports

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented on pkg/rpc/context.go:
```diff
@@ -107,52 +108,118 @@ type connMeta struct {
+var _ base.ModuleTestingKnobs = &ContextTestingKnobs{}
+
+// ContextConfig contains parameters for creating a Context.
+type ContextConfig struct {
```
Given the confusion we have between the terms "context" and "config", a type like `ContextConfig` makes my head spin :) Can it just be `rpc.Config`?

tamird commented:
There are two pretty distinct changes here, and combining them makes the review more difficult. Perhaps you could do the refactoring of exporting whatever heartbeat options you're exporting into a separate commit; then the grpc heartbeat change can be a standalone, less controversial change.

---

Reviewed 6 of 38 files at r1.
Review status: 6 of 38 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/acceptance/util.go, line 278 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFYoGGtXddf_Fro8g:-KgFFYoGGtXddf_Fro8h:b-ez1nle) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/acceptance/util.go#L278)):*
> ```Go
> 		HeartbeatInterval:     time.Second,
> 		HeartbeatTimeout:      time.Second,
> 		EnableClockSkewChecks: true,
> ```

why does this need clock skew checks or heartbeats?

---

*[pkg/testutils/localtestcluster/local_test_cluster.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFq9m2eEXybqQvbzS:-KgFFq9m2eEXybqQvbzT:bmcnwv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/testutils/localtestcluster/local_test_cluster.go#L101)):*
> ```Go
> 		Config:                baseCtx,
> 		HLCClock:              ltc.Clock,
> 		HeartbeatInterval:     time.Second,
> ```

how about exporting a default? by the way, it seems really strange to me that you export these parameters which are not useful to the majority of callers.

---

*[pkg/testutils/testcluster/testcluster_test.go, line 217 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFeJK29q6qtUGdy6q:-KgFFeJK29q6qtUGdy6r:bzi1iub) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/testutils/testcluster/testcluster_test.go#L217)):*
> ```Go
> 		Config:   tc.Server(1).BaseConfig(),
> 		HLCClock: tc.Server(1).Clock(),
> 		// Disable heartbeats and clock skew checks for this connection that's
> ```

is this going to be in every caller?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-:-KgFGMlyHuWVvW-P3wBc:bvyijk8)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 6 of 38 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/rpc/context_test.go, line 594 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgF9Vkoej6HamuSX7Tc:-KgFGwvCjLBZRjWja0aD:b-dmario) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context_test.go#L594)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

s/mojority/majority
</blockquote></details>

lol. could be a new term, "majority required for the cluster to keep its mojo"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
rpc: unexport SourceAddr

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14376)
<!-- Reviewable:end -->

andreimatei commented:
Now that the Keepalive stuff is merged, please take another look at this guy.
The first commit is #14732.

---

Review status: 0 of 43 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/acceptance/util.go, line 278 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFYoGGtXddf_Fro8g:-KgGA7_9WjY56x1rNi5w:b-bctpuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/acceptance/util.go#L278)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why does this need clock skew checks or heartbeats?
</blockquote></details>

Good catch; I think it needs heartbeats (to the extent that anybody needs them), but not skew checks. Added comments.

---

*[pkg/rpc/context.go, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgilSHJ79wV0CvJW-r1-126:-KgGCy5BNrp2KXXnI6Jd:b-2d9acj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context.go#L126)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Given the confusion we have between the terms "context" and "config", a type like `ContextConfig` makes my head spin :) Can it just be `rpc.Config`?
</blockquote></details>

done.
yo dawg I heard you like to config your contexts

---

*[pkg/rpc/context.go, line 138 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgilSHJ79wV0CvJW-r1-138:-KgGDyC_WO7dihPtRRw9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context.go#L138)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"health status of a connection" suggests a lot of overlap with the keepalive. Maybe more accurate to say "the health status of a remote node" instead?
</blockquote></details>

done

---

*[pkg/rpc/context.go, line 216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgilSHJ79wV0CvJW-r1-216:-KgGGQAo-W52kwHu8ol4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context.go#L216)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

this value should be a constant next to defaultKeepaliveInterval
</blockquote></details>

Done.

---

*[pkg/rpc/context.go, line 314 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgilSHJ79wV0CvJW-r1-314:-KgGGZHkdHtWWx9agGd-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context.go#L314)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

remove "if we got a timeout" (it makes this harder to read)
</blockquote></details>

Done.

---

*[pkg/rpc/context_test.go, line 594 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgF9Vkoej6HamuSX7Tc:-KgGGoHkkCmLM2bGxsr-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context_test.go#L594)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

lol. could be a new term, "majority required for the cluster to keep its mojo"
</blockquote></details>

Done.

---

*[pkg/server/server_test.go, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFACB-4r9JSzDY39O9:-KgGGvwQPOSlDPnJG0Cw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/server/server_test.go#L121)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

s/it's goint to report/it reports
</blockquote></details>

Done.

---

*[pkg/testutils/localtestcluster/local_test_cluster.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFq9m2eEXybqQvbzS:-KgGH7gJ19gT4lwyC3Rp:b-z19ea6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/testutils/localtestcluster/local_test_cluster.go#L101)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how about exporting a default? by the way, it seems really strange to me that you export these parameters which are not useful to the majority of callers.
</blockquote></details>

well there is a default - `0` - which means "no heartbeats. I've added another constant in the RPC package and used it in a few places.

---

*[pkg/testutils/testcluster/testcluster_test.go, line 217 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFeJK29q6qtUGdy6q:-KgL-3-kPQfeAp6BM3y5:b-hv947o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/testutils/testcluster/testcluster_test.go#L217)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this going to be in every caller?
</blockquote></details>

yes...

---

*[pkg/testutils/testcluster/testcluster_test.go, line 223 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgimQKbFOMEuR6t3-r1-223:-KgKztT-vqJtpsFV4Hea:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/testutils/testcluster/testcluster_test.go#L223)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this fits in one line now
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-:-KhJulCAcFRPQvhyqmgP:buzdcvc)*
<!-- Sent from Reviewable.io -->

tamird commented:
I'm going to wait for #14732
<https://github.com/cockroachdb/cockroach/pull/14732> before reviewing this.

On Sun, Apr 9, 2017 at 6:53 PM, Andrei Matei <notifications@github.com>
wrote:

> Now that the Keepalive stuff is merged, please take another look at this
> guy.
> The first commit is #14732
> <https://github.com/cockroachdb/cockroach/pull/14732>.
> ------------------------------
>
> Review status: 0 of 43 files reviewed at latest revision, 10 unresolved
> discussions.
> ------------------------------
>
> *pkg/acceptance/util.go, line 278 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFYoGGtXddf_Fro8g:-KgGA7_9WjY56x1rNi5w:b-bctpuw>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/acceptance/util.go#L278>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> why does this need clock skew checks or heartbeats?
>
> Good catch; I think it needs heartbeats (to the extent that anybody needs
> them), but not skew checks. Added comments.
> ------------------------------
>
> *pkg/rpc/context.go, line 126 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgilSHJ79wV0CvJW-r1-126:-KgGCy5BNrp2KXXnI6Jd:b-2d9acj>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context.go#L126>):*
> *Previously, RaduBerinde wrote…*
>
> Given the confusion we have between the terms "context" and "config", a
> type like ContextConfig makes my head spin :) Can it just be rpc.Config?
>
> done.
> yo dawg I heard you like to config your contexts
> ------------------------------
>
> *pkg/rpc/context.go, line 138 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgilSHJ79wV0CvJW-r1-138:-KgGDyC_WO7dihPtRRw9:bb7f2mf>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context.go#L138>):*
> *Previously, RaduBerinde wrote…*
>
> "health status of a connection" suggests a lot of overlap with the
> keepalive. Maybe more accurate to say "the health status of a remote node"
> instead?
>
> done
> ------------------------------
>
> *pkg/rpc/context.go, line 216 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgilSHJ79wV0CvJW-r1-216:-KgGGQAo-W52kwHu8ol4:b-896fix>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context.go#L216>):*
> *Previously, RaduBerinde wrote…*
>
> this value should be a constant next to defaultKeepaliveInterval
>
> Done.
> ------------------------------
>
> *pkg/rpc/context.go, line 314 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgilSHJ79wV0CvJW-r1-314:-KgGGZHkdHtWWx9agGd-:b-896fix>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context.go#L314>):*
> *Previously, RaduBerinde wrote…*
>
> remove "if we got a timeout" (it makes this harder to read)
>
> Done.
> ------------------------------
>
> *pkg/rpc/context_test.go, line 594 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgF9Vkoej6HamuSX7Tc:-KgGGoHkkCmLM2bGxsr-:b-896fix>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/rpc/context_test.go#L594>):*
> *Previously, RaduBerinde wrote…*
>
> lol. could be a new term, "majority required for the cluster to keep its
> mojo"
>
> Done.
> ------------------------------
>
> *pkg/server/server_test.go, line 121 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFACB-4r9JSzDY39O9:-KgGGvwQPOSlDPnJG0Cw:b-896fix>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/server/server_test.go#L121>):*
> *Previously, asubiotto (Alfonso Subiotto Marqués) wrote…*
>
> s/it's goint to report/it reports
>
> Done.
> ------------------------------
>
> *pkg/testutils/localtestcluster/local_test_cluster.go, line 101 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFq9m2eEXybqQvbzS:-KgGH7gJ19gT4lwyC3Rp:b-z19ea6>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/testutils/localtestcluster/local_test_cluster.go#L101>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> how about exporting a default? by the way, it seems really strange to me
> that you export these parameters which are not useful to the majority of
> callers.
>
> well there is a default - 0 - which means "no heartbeats. I've added
> another constant in the RPC package and used it in a few places.
> ------------------------------
>
> *pkg/testutils/testcluster/testcluster_test.go, line 217 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgFFeJK29q6qtUGdy6q:-KgL-3-kPQfeAp6BM3y5:b-hv947o>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/testutils/testcluster/testcluster_test.go#L217>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> is this going to be in every caller?
>
> yes...
> ------------------------------
>
> *pkg/testutils/testcluster/testcluster_test.go, line 223 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KgBTgimQKbFOMEuR6t3-r1-223:-KgKztT-vqJtpsFV4Hea:b-896fix>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/ec364827fa535362dc15c824643eb171407cf7b6/pkg/testutils/testcluster/testcluster_test.go#L223>):*
> *Previously, RaduBerinde wrote…*
>
> [nit] this fits in one line now
>
> Done.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-:-KhJulCAcFRPQvhyqmgP:buzdcvc>*
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14376#issuecomment-292819459>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPE_I3iu7JaKZQaJcjixrIDkcRmmnks5ruWFxgaJpZM4MpmBH>
> .
>

andreimatei commented:
Aight, #14732 has been merged.

---

Review status: 2 of 41 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-:-KhPBU9PaAsN_4r5h_Iu:bwvifca)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 38 files at r1, 4 of 34 files at r2.
Review status: 6 of 41 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/util.go, line 287 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSMRvysI9QD9dsujRB:-KhSMRvysI9QD9dsujRC:bdz1une) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/acceptance/util.go#L287)):*
> 	// We need to configure a MaxOffset on this clock so that the rpc.Context will
> 	// enforce the offset. We're going to initialize the client.Txn using the
> 	// Context's clock and send them through the ExternalSender, so the client's
> 	// clock needs to be synchronized.
> 	//
> 	// TODO(andrei): It's unfortunate that this client, which is not part of the
> 	// cluster, needs to do offset checks. Also, we igore the env variable that
> 	// may control a different acceptable offset for the nodes in the cluster. We
> 	// should stop creating transaction outside of the cluster.

This whole comment seems stale now; or at least it is anchored on the wrong thing. It should probably be down on the `EnableClockSkewChecks: true` line.

---

*[pkg/rpc/clock_offset.go, line 125 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSMsbf_3qkxKlVmyhb:-KhSMsbf_3qkxKlVmyhc:b-cplaoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/clock_offset.go#L125)):*
> ```Go
> ) {
> 	if addr == "127.0.0.1:0" {
> 		log.Fatal(context.TODO(), "uninitialized address in UpdateOffset call")
> ```

panic instead

---

*[pkg/rpc/context.go, line 71 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSNIaxNgLRRBpCSBQ0:-KhSNIayTPuf1sB7Fgym:bsojo2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L71)):*
> ```Go
> )
> 
> var defaultEnableRPCCompression = envutil.EnvOrDefaultBool("COCKROACH_ENABLE_RPC_COMPRESSION", false)
> ```

what is the story with all this movement? did you get carried away?

---

*[pkg/rpc/context.go, line 151 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSNVzjaryiVrub6hWU:-KhSNVzjaryiVrub6hWV:b-ptzolb) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L151)):*
> ```Go
> 	BreakerFactory func() *circuit.Breaker
> 	// PeriodicClockCheckHook is used to override the default clock offset check
> 	// on connections. If an error is returned, the node will crash itself.
> ```

the node will crash itself? what does that even mean in a test?

---

*[pkg/rpc/context.go, line 157 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSN_76yijbN4-pLs2e:-KhSN_76yijbN4-pLs2f:bedhfgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L157)):*
> ```Go
> // ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.
> func (*ContextTestingKnobs) ModuleTestingKnobs() {
> }
> ```

put this on the above line.

---

*[pkg/rpc/context.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSNcqX3Xq8xqNmea_Y:-KhSNcqX3Xq8xqNmea_Z:bb7qx3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L163)):*
> ```Go
> // Config contains parameters for creating a Context.
> type Config struct {
> 	// The base.Config used to open gRPC connections.
> ```

wat?

---

*[pkg/rpc/context.go, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSNkoUuGrDUDJ_hVZr:-KhSNkoUuGrDUDJ_hVZs:b-12eap) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L168)):*
> ```Go
> 	// KeepaliveInterval is the interval at which gRPC is going to perform HTTP2
> 	// pings to check that the connection is still alive.
> 	KeepaliveInterval time.Duration
> ```

this is nuts - grpc already has a config struct for this, why are we duplicating it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Reviving the dormant.

---

Review status: 1 of 42 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/acceptance/util.go, line 287 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSMRvysI9QD9dsujRB:-KjxtXlUf9_cI4pZIoIj:b22gqx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/acceptance/util.go#L287)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> 	// We need to configure a MaxOffset on this clock so that the rpc.Context will
> 	// enforce the offset. We're going to initialize the client.Txn using the
> 	// Context's clock and send them through the ExternalSender, so the client's
> 	// clock needs to be synchronized.
> 	//
> 	// TODO(andrei): It's unfortunate that this client, which is not part of the
> 	// cluster, needs to do offset checks. Also, we igore the env variable that
> 	// may control a different acceptable offset for the nodes in the cluster. We
> 	// should stop creating transaction outside of the cluster.

This whole comment seems stale now; or at least it is anchored on the wrong thing. It should probably be down on the `EnableClockSkewChecks: true` line.
</blockquote></details>

but the comment is primarily about the `MaxOffset` on the line just below. I've reworded it to also refer to the `EnableClockSkewChecks` below, but I'd leave it here.

---

*[pkg/rpc/clock_offset.go, line 125 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSMsbf_3qkxKlVmyhb:-Kjxu1AKcDqh9_u4_DOe:b-9fn7yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/clock_offset.go#L125)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

panic instead
</blockquote></details>

since there's a `ctx` in this method, I made the `log.Fatal` use it.

---

*[pkg/rpc/context.go, line 71 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSNIaxNgLRRBpCSBQ0:-KjxuuDl_XlEFJZy-sMV:b-y0oqlf) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L71)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what is the story with all this movement? did you get carried away?
</blockquote></details>

I moved the old `enableRPCCompression` here now that it comes tied with some constants.

And I've made the compression configurable per `rpc.Context`, since a bunch of other things are becoming configurable. I think it's a good thing...

---

*[pkg/rpc/context.go, line 151 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSNVzjaryiVrub6hWU:-KjxvdB-xhIKsQy1-4mk:ba1af32) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L151)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

the node will crash itself? what does that even mean in a test?
</blockquote></details>

replaced with "panic"

---

*[pkg/rpc/context.go, line 157 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSN_76yijbN4-pLs2e:-KjxvgXo4PgjRHuAU_l7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

put this on the above line.
</blockquote></details>

Done.

---

*[pkg/rpc/context.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSNcqX3Xq8xqNmea_Y:-KjxwKlVjmSgqIJ1pyAH:bfqfsa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L163)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

wat?
</blockquote></details>

I dunno... What would you say about this field?

---

*[pkg/rpc/context.go, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KhSNkoUuGrDUDJ_hVZr:-KjxwxKLTOiOiK49CPMv:bmaantc) ([raw file](https://github.com/cockroachdb/cockroach/blob/55da73577670a5ef1095bb87691268b2bb666505/pkg/rpc/context.go#L168)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is nuts - grpc already has a config struct for this, why are we duplicating it?
</blockquote></details>

ok, switched to using that

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-:-KjyBNRKl2pktyLZQjrf:b-qh24jf)*
<!-- Sent from Reviewable.io -->

tamird commented:
Before I dive into this, could you reorder the commits and remove the body of the "unexport" commit's message? I think that'll alleviate some pain as we continue to iterate.

andreimatei commented:
done

---

Review status: 0 of 42 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-:-KjyGDMp8DOXuu_VVVLG:bb7f2mf)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 41 of 41 files at r3, 39 of 42 files at r4.
Review status: 39 of 42 files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/util.go, line 278 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyGY5CE8GkJq9BoM5p:-KjyGY5CE8GkJq9BoM5q:b-n47bpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/acceptance/util.go#L278)):*
> so that the rpc.Context will enforce
> 	// the offset

does setting the offset do this? i thought setting `EnableClockSkewChecks` did.

---

*[pkg/acceptance/cluster/localcluster.go, line 798 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyGnLRgmISTih1Z4P5:-KjyGnLSwej3MY9BhPwW:b87ht4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/acceptance/cluster/localcluster.go#L798)):*
> ```Go
> 		HeartbeatInterval: rpc.ArbitraryServerHeartbeatInterval,
> 		HeartbeatTimeout:  rpc.ArbitraryServerHeartbeatTimeout,
> 		// No clock skew on local cluster.
> ```

Can you elaborate? These are different containers, which can experience clock skew from one another. The current comment simply repeats the code, which isn't helpful.

Perhaps you meant to say that this client isn't part of the cluster?

---

*[pkg/gossip/client_test.go, line 57 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyMkwEyzmmDC7NGst3:-KjyMkwEyzmmDC7NGst4:b-avskql) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/gossip/client_test.go#L57)):*
> ```Go
> 		HeartbeatInterval:     rpc.ArbitraryServerHeartbeatInterval,
> 		HeartbeatTimeout:      rpc.ArbitraryServerHeartbeatTimeout,
> 		EnableClockSkewChecks: true,
> ```

you have other tests that indicate these checks aren't needed, but then you include them here without a comment. I think you should just enable them everywhere, is there any reason not to?

---

*[pkg/gossip/simulation/network.go, line 79 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyOVNzStCE6B3P5rp_:-KjyOVO-d3fEyHWM9JMc:b-gb0jmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/gossip/simulation/network.go#L79)):*
> ```Go
> 		Config: &base.Config{Insecure: true},
> 		Clock:  hlc.NewClock(hlc.UnixNano, time.Nanosecond),
> 		// Disable heartbeats.
> ```

Seems OK, but give a reason?

---

*[pkg/kv/db_test.go, line 57 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyNMp19PgJ_yAweSn7:-KjyNDLAew-f3_ZWRGua:b-5rbfqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/kv/db_test.go#L57)):*
> ```Go
> 		Config: &ctx,
> 		Clock:  s.Clock(),
> 		// Disable heartbeats for this client that's external to the cluster.
> ```

heh, why does it share a clock with the cluster, then?

---

*[pkg/server/node_test.go, line 72 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyLt5odLqsvRNsnp0g:-KjyLt5odLqsvRNsnp0h:b-blk6rt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/server/node_test.go#L72)):*
> ```Go
> 		Config: nodeTestBaseContext,
> 		Clock:  cfg.Clock,
> 		// Disable heartbeats. Not needed for these tests.
> ```

Do they hurt? If not, perhaps it's better that we leave them?

---

*[pkg/server/server.go, line 165 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyM2J_-6w43sdx0wbn:-KjyM2J_-6w43sdx0wbo:b-upab8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/server/server.go#L165)):*
> ```Go
> 	}
> 
> 	var rpcCtxTestingKnobs rpc.ContextTestingKnobs
> ```

could avoid this by:
```
rpcContextCfg := {
...
}
if knobs := ...; knobs != nil {
  rpcContextCfg.TestingKnobs = *(knobs.(*rpc.ContextTestingKnobs))
```

---

*[pkg/server/testserver.go, line 474 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyMUR_4-yf-8F9UN-G:-KjyMUR_4-yf-8F9UN-H:b-7nqz3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/server/testserver.go#L474)):*
> ```Go
> }
> 
> // BaseConfig is part of TestServerInterface.
> ```

missing "the"

---

*[pkg/sql/distsqlrun/outbox_test.go, line 392 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyMtAlOpu0Sk9VQQP1:-KjyMtAmXuztznB2bpEt:b-9rh5gz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/sql/distsqlrun/outbox_test.go#L392)):*
> ```Go
> 		Config: &base.Config{Insecure: true},
> 		Clock:  hlc.NewClock(hlc.UnixNano, time.Nanosecond),
> 		// Disable heartbeats. Not needed for these tests.
> ```

Does it hurt to enable them?

---

*[pkg/storage/client_test.go, line 123 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyN3oIpUT3IOxv_7JM:-KjyN3oJkVybL8hLGZdh:b-bu5ts8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/storage/client_test.go#L123)):*
> ```Go
> 		Config: &base.Config{Insecure: true},
> 		Clock:  storeCfg.Clock,
> 		// Disable heartbeats. Not needed for these tests.
> ```

Same comment as elsewhere.

---

*[pkg/storage/client_test.go, line 284 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyN5YH2M0xu3dsfU5U:-KjyN5YH2M0xu3dsfU5V:b-bu5ts8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/storage/client_test.go#L284)):*
> ```Go
> 			Config: &base.Config{Insecure: true},
> 			Clock:  m.clock,
> 			// Disable heartbeats. Not needed for these tests.
> ```

Same comment as elsewhere.

---

*[pkg/storage/raft_transport_test.go, line 117 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyN1P5-xMiqmGAKrol:-KjyN1P5-xMiqmGAKrom:b-bu5ts8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/storage/raft_transport_test.go#L117)):*
> ```Go
> 		Config: testutils.NewNodeTestBaseConfig(),
> 		Clock:  hlc.NewClock(hlc.UnixNano, time.Nanosecond),
> 		// Disable heartbeats. Not needed for these tests.
> ```

Same comment as elsewhere.

---

*[pkg/testutils/testcluster/testcluster_test.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376#-KjyNtyX5Ii6daSVPztZ:-KjyNtyX5Ii6daSVPzt_:b-5rbfqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e403568b83c96e59b2c42c27874ca4b80537a6d7/pkg/testutils/testcluster/testcluster_test.go#L217)):*
> ```Go
> 		Config: tc.Server(1).BaseConfig(),
> 		Clock:  tc.Server(1).Clock(),
> 		// Disable heartbeats and clock skew checks for this connection that's
> ```

heh, why does it share a clock with the cluster, then?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14376)*
<!-- Sent from Reviewable.io -->

tamird commented:
@andreimatei do you plan to return to this?

andreimatei commented:
Yes

On Oct 22, 2017 2:16 PM, "Tamir Duberstein" <notifications@github.com>
wrote:

> @andreimatei <https://github.com/andreimatei> do you plan to return to
> this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14376#issuecomment-338497696>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcR1ZbMfQk7m2doi8V0U3yy3Uc8ccks5su4aKgaJpZM4MpmBH>
> .
>

andreimatei reviewed (commented):
well, I've returned to this in that I've opened #33701 to try to get momentum again

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14376#-:-LW7obhP-m_ikJySs9rm:bqmwbms)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=14376) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=14376) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=14376) it.</sub>
